src/software/unix_full_system_main.cpp

```diff
@@ -62,14 +91,40 @@ int main(int argc, char** argv)
             TracySetProgramName("Thunderbots: Yellow");
         }
 
-        LoggerSingleton::initializeLogger(args.runtime_dir);
+        std::function<double()> time_provider;
+        if (!args.ci)
+        {
+            // Return the current time since epoch in seconds
+            time_provider = []() {
+                return std::chrono::duration<double>(
+                           std::chrono::system_clock::now().time_since_epoch())
+                    .count();
+            };
+        }
+        else
+        {
+            // Default to returning 0 and update the time_provider with the World
+            // timestamp once the backend is set up
+            time_provider = []() { return 0; };
+        }
+        proto_logger = std::make_shared<ProtoLogger>(args.runtime_dir, time_provider,
+                                                     args.friendly_colour_yellow);
+        LoggerSingleton::initializeLogger(args.runtime_dir, proto_logger);
         TbotsProto::ThunderbotsConfig tbots_proto;
 
         // Override friendly color
         tbots_proto.mutable_sensor_fusion_config()->set_friendly_color_yellow(
             args.friendly_colour_yellow);
 
-        auto backend = std::make_shared<UnixSimulatorBackend>(args.runtime_dir);
+        auto backend =
+            std::make_shared<UnixSimulatorBackend>(args.runtime_dir, proto_logger);
+        if (args.ci)
+        {
+            // Update the time provider for ProtoLogger
+            proto_logger->updateTimeProvider(
+                [&backend]() { return backend->getLastWorldTimeSec(); });
+        }
```

Comment:
is there a chance of race conditions where backend would get the wrong time from the time provider?

Reply:
I guess probably not because the observers are registered later on in this function